ipn/ipnlocal/profiles.go

```diff
@@ -0,0 +1,467 @@
+	}
+	pm.prefs = prefs
+	pm.currentProfile = kp
+	pm.isNewProfile = false
+	return pm.setAsUserSelectedProfileLocked()
+}
+
+func (pm *ProfileManager) setAsUserSelectedProfileLocked() error {
+	k := ipn.CurrentProfileKey(pm.currentUserID)
+	if pm.currentProfile == nil {
+		return pm.store.WriteState(k, nil)
+	}
+	return pm.store.WriteState(k, []byte(pm.currentProfile.Key))
+}
+
+func (pm *ProfileManager) loadSavedPrefs(key ipn.StateKey) (ipn.PrefsView, error) {
+	bs, err := pm.store.ReadState(key)
+	if err != nil {
+		if err == ipn.ErrStateNotExist {
+			return emptyPrefs, nil
+		}
+		return ipn.PrefsView{}, err
+	}
+	savedPrefs, err := ipn.PrefsFromBytes(bs)
+	if err != nil {
+		return ipn.PrefsView{}, fmt.Errorf("PrefsFromBytes: %v", err)
+	}
+	pm.logf("using backend prefs for %q: %v", key, savedPrefs.Pretty())
+
+	// Ignore any old stored preferences for https://login.tailscale.com
+	// as the control server that would override the new default of
+	// controlplane.tailscale.com.
+	if savedPrefs.ControlURL != "" &&
+		savedPrefs.ControlURL != ipn.DefaultControlURL &&
+		ipn.IsLoginServerSynonym(savedPrefs.ControlURL) {
+		savedPrefs.ControlURL = ""
+	}
+	return savedPrefs.View(), nil
+}
+
+// CurrentProfile returns the name and ID of the current profile, or "" if the profile
+// is not named.
+func (pm *ProfileManager) CurrentProfile() ipn.LoginProfile {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	if pm.currentProfile == nil {
+		return ipn.LoginProfile{}
+	}
+	return *pm.currentProfile
+}
+
+var ErrProfileNotFound = errors.New("profile not found")
+
+// DeleteProfile removes the profile with the given id. It is a no-op if the
+// profile does not exist.
+func (pm *ProfileManager) DeleteProfile(profile ipn.ProfileID) error {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	kp, ok := pm.knownProfiles[profile]
+	if !ok {
+		return ErrProfileNotFound
+	}
+	if kp.ID == pm.currentProfile.ID {
+		pm.NewProfile()
+	}
+	if err := pm.store.WriteState(kp.Key, nil); err != nil {
+		return err
+	}
+	delete(pm.knownProfiles, profile)
+	return pm.writeKnownProfiles()
+}
+
+func (pm *ProfileManager) writeKnownProfiles() error {
+	b, err := json.Marshal(pm.knownProfiles)
+	if err != nil {
+		return err
+	}
+	return pm.store.WriteState(ipn.KnownProfilesStateKey, b)
+}
+
+// NewProfile creates a new profile with the given name. It switches to the new
+// profile. The new profile is not persisted until SetPrefs is called.
+func (pm *ProfileManager) NewProfile() {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	pm.prefs = emptyPrefs
+	pm.currentProfile = nil
+	pm.isNewProfile = true
+}
+
+// emptyPrefs is the default prefs for a new profile.
+var emptyPrefs = func() ipn.PrefsView {
+	prefs := ipn.NewPrefs()
+	prefs.WantRunning = false
+	return prefs.View()
+}()
+
+// Store returns the StateStore used by the ProfileManager.
+func (pm *ProfileManager) Store() ipn.StateStore {
+	return pm.store
+}
+
+func (pm *ProfileManager) ReadProfile(p ipn.ProfileID) (ipn.PrefsView, error) {
```

Comment:
Needs docs (but is also unused)?

Reply:
Oh I added this when I thought we would need to pull Persist info out of the prefs and then decided to store them in the LoginProfile instead